langlib/lang.value/src/main/java/org/ballerinalang/langlib/value/FromJsonWithType.java

```diff
@@ -0,0 +1,291 @@
+import static org.ballerinalang.util.BLangCompilerConstants.VALUE_VERSION;
+
+/**
+ * Extern function lang.values:fromJsonWithType.
+ *
+ * @since 2.0
+ */
+@BallerinaFunction(
+        orgName = "ballerina",
+        packageName = "lang.value", version = VALUE_VERSION,
+        functionName = "fromJsonWithType",
+        args = {
+                @Argument(name = "v", type = TypeKind.ANYDATA),
+                @Argument(name = "t", type = TypeKind.TYPEDESC)
+        },
+        returnType = {
+                @ReturnType(type = TypeKind.ANYDATA),
+                @ReturnType(type = TypeKind.ERROR)
+        },
+        isPublic = true
+)
+public class FromJsonWithType {
+    private static final String AMBIGUOUS_TARGET = "ambiguous target type";
+
+    public static Object fromJsonWithType(Strand strand, Object v, TypedescValue t) {
+        BType describingType = t.getDescribingType();
+        // typedesc<json>.constructFrom like usage
+        if (describingType.getTag() == TypeTags.TYPEDESC_TAG) {
+            return convert(((BTypedescType) t.getDescribingType()).getConstraint(), v, t, strand);
+        }
+        // json.constructFrom like usage
+        return convert(describingType, v, t, strand);
+    }
+
+    public static Object convert(BType convertType, Object inputValue) {
+        return convert(convertType, inputValue, null, null);
+    }
+
+    public static Object convert(BType convertType, Object inputValue, TypedescValue t, Strand strand) {
+        try {
+            return convert(inputValue, convertType, new ArrayList<>(), t, strand);
+        } catch (ErrorValue e) {
+            return e;
+        } catch (BallerinaException e) {
+            return createError(CONSTRUCT_FROM_CONVERSION_ERROR, e.getDetail());
+        }
+    }
+
+    private static Object convert(Object value, BType targetType, List<TypeValuePair> unresolvedValues,
+                                  TypedescValue t, Strand strand) {
+        return convert(value, targetType, unresolvedValues, false, t, strand);
+    }
+
+
+    private static Object convert(Object value, BType targetType, List<TypeValuePair> unresolvedValues,
+                                  boolean allowAmbiguity, TypedescValue t, Strand strand) {
+        if (value == null) {
+            if (targetType.isNilable()) {
+                return null;
+            }
+            return createError(CONSTRUCT_FROM_CONVERSION_ERROR,
+                    BLangExceptionHelper.getErrorMessage(RuntimeErrors.CANNOT_CONVERT_NIL, targetType));
+        }
+        List<BType> convertibleTypes;
+        convertibleTypes = TypeConverter.getConvertibleTypes(value, targetType);
+        if (convertibleTypes.isEmpty()) {
+            if (TypeChecker.getType(value).getTag() == TypeTags.STRING_TAG &&
+                    (targetType.getTag() >= TypeTags.XML_ELEMENT_TAG || targetType.getTag() == TypeTags.XML_TAG)) {
```

Comment:
Shall we add an upper bound for the subtypes as well for the following? If we add another `anydata` type, this will break.
```
targetType.getTag() >= TypeTags.XML_ELEMENT_TAG
```

Reply:
okay